site/_posts/2018-02-12-discovering-a-ruby-memory-leak.html.md

```diff
@@ -0,0 +1,441 @@
+leaking as well, though, it took much more time for this to be truly observable
+under normal conditions (like a month).  But, since it had a more predictable
+workflow, this was the better candidate to do some deeper process of
+elimination testing.  It took some time to get a scenario where this was
+leaking consistently, but once we did, it was simpler to do A/B testing with
+it.
+
+
+Step 1.5:  Reproducing the issue... fasterer
+--------------------------------------------
+
+One problem with using the `MiqServer` process as the target for observing the
+leak is it took _**days**_/_**weeks**_ to determine if it was leaking or just
+the process taking it's sweet time time to stabilize it's memory usage.  This
+led to very long gaps between making a change and having observable results to
+view from it.
+
+One of the least invasive things that we eventually did was decrease the server
+`:monitor_poll` and `:worker_monitor_frequency` (SPOILER! more on the second
+one in a bit) values from `"5.seconds"` and `"15.seconds"` respectively to
+`"1.second"` for them both.  This decreased the wait time between each
+iteration of it's normal work and, fortunately, increased the leak rate.
+
+We also increased the frequency of normal method calls by doing things like
+`40.times { do_some_work }` in places where we thought the leak might be, and
+if there was rapid growth, we knew where to probe deeper.
+
+These are things that happened throughout the process of discovering the cause
+of the leak, and didn't necessarily happen in between this step and the next
+(hence the "Step 1.5").  Steps to accelerate the feedback loop, such as this,
+should be done as early into the memory leak debugging process as possible.  We
+just weren't so lucky...
+
+
+Step 2:  Narrowing down the culprits
+------------------------------------
+
+As mentioned before, the `MiqServer` process has a pretty regular work loop.
+In case you aren't familiar with the workflow of the `MiqServer` work
+loop, here is a gist of the important steps:
+
+1. heartbeat (tell other servers that I am still alive)
+2. process any messages sent to the server process over the message queue
+3. check on the health status of other server processes (and log them)
+4. check the status of all of the workers processes (stop/start them as needed)
+5. fetch queue work to assign to worker processes
+
+Each of these steps happen in varying intervals, but mostly happen on a
+consistent basis.  Also if everything is running "normally", the work that is
+done is also fairly consistent and repeatable across each one of those steps.
+
+After trying to instrument the main loop with tools like `MemoryProfiler`,
+`stackprof`, and even just looking at regular ruby memory dumps using
+`ObjectSpace.dump_all`, we ended up taking a more low tech approach of just
+removing code by commenting out the steps above, and seeing if the leak still
+happened after a server restart.
```

Comment:
Split in to two sentences. "We first tried to instrument the main loop...and even just looking at regular Ruby memory dumps...", then "In the end, we took a more low tech approach..."

Reply:
Edit can be found here:

https://gist.github.com/NickLaMuro/730b252a48019cc6bef85270ac16ff44/revisions#diff-4756561d24941a282634640442f5d8be